### ISSUE TITLE
Different training loss results with just changing parallel topology

### ISSUE THREAD

**Issue by jasonjunweilyu** [ecmwf/anemoi-core#1403]
### What happened?

When fine-tuning Bris deterministic model (stretched-grid) with BARRA reanalysis over Tasmania, Australia, I got different loss trajectory and convergence results with just different parallel topology during training (i.e. whether model sharding is used) while keeping the same global learning rate and global batch size for one optimizer update with `accum_grad_batches`.  

Below I show the two runs for their config parameter differences tracked by mlflow. They all use 16 H200 GPUs (4 nodes, each with 4 H200 GPUs). `num_chunks` is changed to decrease peak GPU memory needed:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/JLYU~1.INT/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/JLYU~1.INT/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">

config.training.run_id | 459927887d1a453da66413520bdbea64 | e9032dd79ccb47c2906646d0a5d3320e
-- | -- | --
num_gpus_per_model | 4 | 1
num_chunks   (decoder/processor/encoder) | 4 | 16
config.training.accum_grad_batches | 8 | 2
config.training.lr.rate | 1.00E-06 | 2.50E-07
step 0   train_mse_loss | 0.016867 | 0.035296
final   train_mse_loss_epoch | 0.005 | 0.017
inference   results | Good | Bad

</body>

</html>

Comparing these two runs, you can see that the starting and final training losses are quite different, which seems to suggest different optimisation trajectories. These lead to `459927*` having much better inference results with small global skill degradation using the fine-tuned checkpoint while `e9032d*` has significant global skill degradation, as is shown below.

<img width="1155" height="682" alt="Image" src="https://github.com/user-attachments/assets/04f7ef80-f0a1-42f6-ab69-3c3770266972" />

`459927*` loss and lr curve:

<img width="1611" height="467" alt="Image" src="https://github.com/user-attachments/assets/36dc076a-9d77-4977-aae9-d3ce28b746a5" />

`e9032d*` loss and lr curve:

<img width="1607" height="465" alt="Image" src="https://github.com/user-attachments/assets/cb487fab-919f-46e8-b1f3-350e66f1f51e" />

So far I have tried changing these config parameters to see if they are the cause, but found that they are not:
(1) precision: I have a run `18801f23528e4fb0b2bfec4a80a78f77` that just changed `config.training.precision` in `e9032d*` from `16-mixed` to be `32`, which doesn't improve the results.
(2) number of shared models: I have a run `7546d34fc7be4a75b9efed495eac1884` that just used 4 GPUs in total, compared to `e9032d*`. The `config.training.lr.rate` is changed to `1e-6`. The `config.training.accum_grad_batches` is changed to `8`. The intention of this run is to have the same number of model groups as `459927*`. However, the results of `7546d3*` is similar to `e9032d*`.
(3) `ANEMOI_BASE_SEED`: I want to see whether the seeding is the cause. So I have a run `1cdc1fafdd6d4a52b8e8e106ad8d3372` that just changed `ANEMOI_BASE_SEED` from `1000` to `23456` from `459927*`, and it turned out that the results are very similar.
(4) `keep_batch_sharded`: I have a run `d60e063f38d548acb4e5aa34cd195b0a` that just changed `config.model.keep_batch_sharded` from `True` to `False` from `459927*`. The results are similar.

### What are the steps to reproduce the bug?

1. Use Met-Norway Bris deterministic model checkpoint on HuggingFace
2. Finetune with stretch-grid and have Tasmania ([-40.0, 144.0, -44.0, 149.0]) as the cutout area with BARRA data and rest of the globe using IFS.
3. Do only no-rollout training runs with 16 H200 GPUs (4 nodes, each with 4 H200 GPUs) 

<details>
  <summary>📎 Click to view config for 459927* </summary>

[459927_inference_ckpt_metadata.yaml](https://github.com/user-attachments/files/32273584/459927_inference_ckpt_metadata.yaml)

</details>

<details>
  <summary>📎 Click to view config for e9032d* </summary>

[e9032d_inference_ckpt_metadata.yaml](https://github.com/user-attachments/files/32273632/e9032d_inference_ckpt_metadata.yaml)

</details>

### Version

anemoi-models 0.9.5; anemoi-training 0.6.5; anemoi-graphs 0.6.6; anemoi-datasets 0.5.29; anemoi-transform 0.1.17; anemoi-inference 0.7.3

### Platform (OS and architecture)

NCI-Gadi x86_64 GNU/Linux

### Relevant log output

```shell

```

### Organisation

Bureau of Meteorology, Australia

**Comment by gabrieloks** [comment]
Hello! Did you use MlFlow to track these two runs? Could you share the mlflow data ?

**Comment by ssmmnn11** [comment]
the graph files are identical?

**Comment by jasonjunweilyu** [comment]
Hi, Thanks for the quick responses. The mlflow data for the two runs are here:

[459927_mlflow.zip](https://github.com/user-attachments/files/32284322/459927_mlflow.zip)
[e9032d_mlflow.zip](https://github.com/user-attachments/files/32284323/e9032d_mlflow.zip)

For the graph, I used the same config options for generating graph during training, as can be seen in the mlflow tracked config options or the config in the checkpoint metadata. But since `config.graph.overwrite=True`, the graph is not guaranteed to be identical. Is there a good way to check if there is any difference between the two graphs?

**Comment by jasonjunweilyu** [comment]
I just got one new run `fc3b0eae56db4158ab5874f47fa5c81a` with 16 GPUs in total, and 1 model instance per 2 GPUs. Surprisingly, with model sharding this time, the loss results are bad and similar to `e9032d*`. The mesh is using the same one as `459927*`. So it looks like the mesh is not the cause. 



<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/JLYU~1.INT/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/JLYU~1.INT/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


config.training.run_id | fc3b0eae56db4158ab5874f47fa5c81a
-- | --
num_gpus_per_model | 2
num_chunks   (decoder/processor/encoder) | 16
config.training.accum_grad_batches | 3
config.training.lr.rate | 5.00E-07
step 0   train_mse_loss | 0.036557
final   train_mse_loss_epoch | 0.014264



</body>

</html>

The mlflow data for this run is here: 

[fc3b0e_mlflow.zip](https://github.com/user-attachments/files/32324992/fc3b0e_mlflow.zip)

**Comment by jasonjunweilyu** [comment]
It looks like `num_chunks  (decoder/processor/encoder)` may be the cause. I have another new run `2ce653131f9f462887d4488ac4727b89` which just changed num_chunks from `459927*` but had bad loss results:


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/JLYU~1.INT/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/JLYU~1.INT/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


config.training.run_id | 2ce653131f9f462887d4488ac4727b89
-- | --
num_gpus_per_model | 4
num_chunks   (decoder/processor/encoder) | 16
config.training.accum_grad_batches | 8
config.training.lr.rate | 1.00E-06
step 0   train_mse_loss | 0.038334
final   train_mse_loss_epoch | 0.01697



</body>

</html>


The mlflow data for this run is here:

[2ce653_mlflow.zip](https://github.com/user-attachments/files/32364012/2ce653_mlflow.zip)

It is unclear to me, though, why `num_chunks` would cause differences in loss results.

**Comment by japols** [comment]
Hi, I've noticed you're on a pretty old release of anemoi-models. We had one issue with processor chunking in the past that was addressed [here](https://github.com/ecmwf/anemoi-core/commit/06e5533f3e8da37c44d887c42b67440b40286cb3). This seems to be in line with what you're describing i.e. continuing from a checkpoint, overwriting num_chunks in the processor which leads to randomly initialised processor weights -> bad model.

If you cherry-pick https://github.com/ecmwf/anemoi-core/commit/06e5533f3e8da37c44d887c42b67440b40286cb3, anemoi-training should automatically apply the chunking migration fix to your checkpoint.

**Comment by jasonjunweilyu** [comment]
Hi, thanks a lot for the advice. Indeed my problem seems to align well with the processor chunking problem. I tried a config with just changing the processor `num_chunks` from 4 to 16 and still got bad loss convergence results. I am now cherry-picking the processor chunk fix commit and rerunning this config to confirm this.
